### PR TITLE
Merge JS2 to Main

### PR DESCRIPTION
**Pull request by pretzL**

### CODE REVIEW

**Comment by Moesera** [comment]
# Peer Review

## JS / Functionality
### Landing page / profile page
The profile landing page looks nice, you get a feed of all the posts.
on the profile page the form doesnt work and it only has body, but i guess it is intended.

**What works fine.**
- posting comments works fine.
- View a post's comments looks nice.
- liking post in view post works fine.
- updating banner works great.

**What does not work fine**

- Would enjoy a button to bring me back from view post comments.
- The error message for the avatar update is bit low, couldnt se it when i was editing until i went further down.
- Honestly i dont know what the clear button should do, i expect it to clear the input but doesnt, would just remove it for now if it should not be usable tbh.

### Homepage

**What works fine.**
- editing a post works fine, and i really like the setup
- creating post works fine, other than what i stated down below.
- deleting post works great
- i love your sorting functionality.
- search functionality works fine.

**What does not work fine**
- When i made a post, i didnt get the title back, i only got the body image and tags and can only view body image and tags from others.

### in general
The site functions well and i like alot of the functionality, the code looks good, and the folder structure is well structured in my opinion.

What i am missing is more JsDocs documentation, it is easier to hover over functions and see what gets passed in or what they do and how they work, its easier to manage the code and understand it faster with it. Would recommend to take a look at it, it is also a criteria.
Also when you come back to the code a bit later its easier to get back into your code.

## HTML / Look

- The avatar image scales out of place on the profile page.
- The most buttons i hover over doesnt become a pointer, i would go over the buttons just because it take's away the feel of the website.
- You could also take away the login button when people are logged in at home and my profile page, just doesnt feel right to have it there.
- On smaller screens the hamburger menu does not work.. and you cant navigate because of this.

## in general

The site looks good, and there is not alot of issues. But i would definetly look at the size of the login background image it is 2 MB, you should try to get it down to 200kb max 300.
The image placeholder banner is a bit large too and is around 4000kb.

**The issue:** 
This will cause slow loading speed.

For the image issue i recommend to use the **image** tab then **image size** or **bildestørrelse** in photoshop and shrink it a bit.
And then use the export to web and take down the quality if needed.


### Overall the site looks good, and good luck on the assignment !.

**Comment by Fermain** [comment]
image

Unable to edit post if there is no media present due to validator

File: src/js/posts/createComment.mjs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    },
+  })
+    .then((response) => response.json())
+    .then((json) => {
+      if (json.message) {
+        errorContainer.innerHTML = errorMessage(json.message);
+      } else {
+        errorContainer.innerHTML = successMessage("Post creation");
+
+        setTimeout(() => {
```
Comment (Fermain) [inline src/js/posts/createComment.mjs]: This is a sign that async/await is not being used properly. Check the `teacher-review` branch for my solution to this.

File: src/js/posts/createPost.mjs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    },
+  })
+    .then((response) => response.json())
+    .then((json) => {
+      if (json.message) {
+        errorContainer.innerHTML = errorMessage(json.message);
+      } else {
+        errorContainer.innerHTML = successMessage("Post creation");
+
+        setTimeout(() => {
```
Comment (Fermain) [inline src/js/posts/createPost.mjs]: Here too

File: src/js/posts/createReactions.mjs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    },
+  })
+    .then((response) => response.json())
+    .then((json) => {
+      if (json.message) {
+        errorContainer.innerHTML = errorMessage(json.message);
+      } else {
+        errorContainer.innerHTML = successMessage("Like");
+
+        setTimeout(() => {
```
Comment (Fermain) [inline src/js/posts/createReactions.mjs]: And here

File: src/js/posts/editPost.mjs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    },
+  })
+    .then((response) => response.json())
+    .then((json) => {
+      if (json.message) {
+        errorContainer.innerHTML = errorMessage(json.message);
+      } else {
+        errorContainer.innerHTML = successMessage("Post edit");
+
+        setTimeout(() => {
```
Comment (Fermain) [inline src/js/posts/editPost.mjs]: Here too!

File: src/js/posts/getPostToDelete.mjs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+  deletePostButton.forEach((e) => {
+    e.addEventListener("click", (i) => {
+      const closestPost = i.target.closest(".card-body");
+      const postId = closestPost.childNodes[3].innerText;
+
+      deletePost(postId);
+
+      // Reloads the website after 1 second after clicking delete.
+      setTimeout(location.reload.bind(location), 1000);
```
Comment (Fermain) [inline src/js/posts/getPostToDelete.mjs]: You guessed it

File: src/js/user/profiles/follow-unfollow/requests.mjs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    },
+  })
+    .then((response) => response.json())
+    .then((json) => {
+      if (json.message) {
+        errorContainer.innerHTML = errorMessage(json.message);
+      } else {
+        errorContainer.innerHTML = successMessage("Post creation");
+
+        setTimeout(() => {
```
Comment (Fermain) [inline src/js/user/profiles/follow-unfollow/requests.mjs]: Hello darkness my old friend

File: src/js/user/profiles/updateProfile.mjs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    },
+  })
+    .then((response) => response.json())
+    .then((json) => {
+      if (json.message) {
+        errorContainer.innerHTML = errorMessage(json.message);
+      } else {
+        errorContainer.innerHTML = successMessage("Profile media edit");
+
+        setTimeout(() => {
```
Comment (Fermain) [inline src/js/user/profiles/updateProfile.mjs]: I've come to talk with you again

File: src/js/user/registerUser.mjs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      },
+    })
+      .then((response) => response.json)
+      .then((json) => {
+        if (json.error) {
+          errorContainer.innerHTML = errorMessage(json.message);
+        } else {
+          errorContainer.innerHTML = successMessage("Registration");
+
+          setTimeout(() => {
```
Comment (Fermain) [inline src/js/user/registerUser.mjs]: Because a vision softly creeping,
Left its seeds while I was sleeping

**Review by Fermain — CHANGES_REQUESTED** [review]
This is a really high quality submission that looks and feels like a real application. There are some issues to be resolved (setTimeout mostly) and there are improvements to be made in terms of code organisation and structure. (await > then).

With the setTimeout issue and update post issues solved this is a very, very comfortable pass.

**Review by Fermain — CHANGES_REQUESTED** [review]
It looks like you have some artefacts left over from installing a bundler (out directory).

If these files aren't being used they should be removed from history. It's conventional when using a bundler to ignore the output directory so that this content is not included in a pull request.

Otherwise, the code is working great.

**Pull request merged**